File: matrix.go
// matrix.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Task struct {
	ID          int       `json:"id"`
	Title       string    `json:"title"`
	Description string    `json:"description"`
	Important   bool      `json:"important"`
	Urgent      bool      `json:"urgent"`
	Done        bool      `json:"done"`
	CreatedAt   time.Time `json:"created_at"`
}

func (t Task) Quadrant() string {
	if t.Important && t.Urgent {
		return "Важно и срочно"
	} else if t.Important && !t.Urgent {
		return "Важно, но не срочно"
	} else if !t.Important && t.Urgent {
		return "Не важно, но срочно"
	} else {
		return "Не важно и не срочно"
	}
}

type EisenhowerMatrix struct {
	tasks   []Task
	nextID  int
}

func NewEisenhowerMatrix() *EisenhowerMatrix {
	return &EisenhowerMatrix{
		tasks:  []Task{},
		nextID: 1,
	}
}

func (m *EisenhowerMatrix) AddTask(title, desc string, important, urgent bool) Task {
	task := Task{
		ID:          m.nextID,
		Title:       title,
		Description: desc,
		Important:   important,
		Urgent:      urgent,
		Done:        false,
		CreatedAt:   time.Now(),
	}
	m.tasks = append(m.tasks, task)
	m.nextID++
	return task
}

func (m *EisenhowerMatrix) FindTask(id int) *Task {
	for i := range m.tasks {
		if m.tasks[i].ID == id {
			return &m.tasks[i]
		}
	}
	return nil
}

func (m *EisenhowerMatrix) EditTask(id int, title, desc *string, important, urgent *bool) bool {
	task := m.FindTask(id)
	if task == nil {
		return false
	}
	if title != nil {
		task.Title = *title
	}
	if desc != nil {
		task.Description = *desc
	}
	if important != nil {
		task.Important = *important
	}
	if urgent != nil {
		task.Urgent = *urgent
	}
	return true
}

func (m *EisenhowerMatrix) ToggleDone(id int) bool {
	task := m.FindTask(id)
	if task == nil {
		return false
	}
	task.Done = !task.Done
	return true
}

func (m *EisenhowerMatrix) DeleteTask(id int) bool {
	for i, t := range m.tasks {
		if t.ID == id {
			m.tasks = append(m.tasks[:i], m.tasks[i+1:]...)
			return true
		}
	}
	return false
}

func (m *EisenhowerMatrix) FilterByQuadrant(quadrant string) []Task {
	var result []Task
	for _, t := range m.tasks {
		if t.Quadrant() == quadrant {
			result = append(result, t)
		}
	}
	return result
}

func (m *EisenhowerMatrix) Stats() map[string]interface{} {
	total := len(m.tasks)
	doneCount := 0
	quadCounts := make(map[string]int)
	for _, t := range m.tasks {
		if t.Done {
			doneCount++
		}
		quadCounts[t.Quadrant()]++
	}
	return map[string]interface{}{
		"total":     total,
		"done":      doneCount,
		"pending":   total - doneCount,
		"quadrants": quadCounts,
	}
}

func (m *EisenhowerMatrix) SaveToFile(filename string) error {
	data, err := json.MarshalIndent(m.tasks, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filename, data, 0644)
}

func (m *EisenhowerMatrix) LoadFromFile(filename string) error {
	data, err := os.ReadFile(filename)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	var loaded []Task
	if err := json.Unmarshal(data, &loaded); err != nil {
		return err
	}
	m.tasks = loaded
	for _, t := range loaded {
		if t.ID >= m.nextID {
			m.nextID = t.ID + 1
		}
	}
	return nil
}

func readString(prompt string) string {
	fmt.Print(prompt)
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	return strings.TrimSpace(input)
}

func readBool(prompt string) bool {
	for {
		input := readString(prompt)
		if input == "" {
			return false
		}
		if input == "y" || input == "Y" {
			return true
		} else if input == "n" || input == "N" {
			return false
		}
		fmt.Println("Введите y или n.")
	}
}

func readInt(prompt string) int {
	for {
		input := readString(prompt)
		if val, err := strconv.Atoi(input); err == nil {
			return val
		}
		fmt.Println("Введите число.")
	}
}

func printTask(task Task, showQuadrant bool) {
	doneMark := "✅"
	if !task.Done {
		doneMark = "⬜"
	}
	fmt.Printf("%s #%d - %s\n", doneMark, task.ID, task.Title)
	fmt.Printf("   Описание: %s\n", task.Description)
	if showQuadrant {
		fmt.Printf("   Квадрант: %s (важно: %t, срочно: %t)\n", task.Quadrant(), task.Important, task.Urgent)
	}
	fmt.Printf("   Создана: %s\n", task.CreatedAt.Format("2006-01-02 15:04:05"))
}

func main() {
	matrix := NewEisenhowerMatrix()
	if err := matrix.LoadFromFile("eisenhower_data.json"); err != nil {
		fmt.Println("Ошибка загрузки:", err)
	}

	for {
		fmt.Println("\n===== МАТРИЦА ЭЙЗЕНХАУЭРА (Go) =====")
		fmt.Println("1. Показать все задачи")
		fmt.Println("2. Добавить задачу")
		fmt.Println("3. Редактировать задачу")
		fmt.Println("4. Отметить выполненной / невыполненной")
		fmt.Println("5. Удалить задачу")
		fmt.Println("6. Показать статистику")
		fmt.Println("7. Сохранить в файл")
		fmt.Println("8. Загрузить из файла")
		fmt.Println("0. Выход")
		choice := readString("Выберите действие: ")

		switch choice {
		case "0":
			return
		case "1":
			if len(matrix.tasks) == 0 {
				fmt.Println("Задач нет.")
			} else {
				quadrants := []string{"Важно и срочно", "Важно, но не срочно", "Не важно, но срочно", "Не важно и не срочно"}
				for _, q := range quadrants {
					tasks := matrix.FilterByQuadrant(q)
					if len(tasks) > 0 {
						fmt.Printf("\n--- %s ---\n", q)
						for _, t := range tasks {
							printTask(t, false)
						}
					}
				}
			}
		case "2":
			title := readString("Введите название: ")
			if title == "" {
				fmt.Println("Название не может быть пустым.")
				continue
			}
			desc := readString("Введите описание: ")
			important := readBool("Задача важна? (y/n): ")
			urgent := readBool("Задача срочна? (y/n): ")
			task := matrix.AddTask(title, desc, important, urgent)
			fmt.Printf("Задача добавлена с ID %d, квадрант: %s\n", task.ID, task.Quadrant())
		case "3":
			id := readInt("Введите ID задачи: ")
			task := matrix.FindTask(id)
			if task == nil {
				fmt.Println("Задача не найдена.")
				continue
			}
			fmt.Println("Оставьте поле пустым, чтобы не менять.")
			newTitle := readString(fmt.Sprintf("Новое название (%s): ", task.Title))
			if newTitle != "" {
				task.Title = newTitle
			}
			newDesc := readString(fmt.Sprintf("Новое описание (%s): ", task.Description))
			if newDesc != "" {
				task.Description = newDesc
			}
			newImp := readString(fmt.Sprintf("Важна? (y/n) сейчас: %t: ", task.Important))
			if newImp != "" {
				task.Important = newImp == "y" || newImp == "Y"
			}
			newUrg := readString(fmt.Sprintf("Срочна? (y/n) сейчас: %t: ", task.Urgent))
			if newUrg != "" {
				task.Urgent = newUrg == "y" || newUrg == "Y"
			}
			fmt.Println("Задача обновлена.")
		case "4":
			id := readInt("Введите ID задачи: ")
			if matrix.ToggleDone(id) {
				fmt.Println("Статус выполнения изменён.")
			} else {
				fmt.Println("Задача не найдена.")
			}
		case "5":
			id := readInt("Введите ID задачи для удаления: ")
			if matrix.DeleteTask(id) {
				fmt.Println("Задача удалена.")
			} else {
				fmt.Println("Задача не найдена.")
			}
		case "6":
			stats := matrix.Stats()
			fmt.Println("\n=== СТАТИСТИКА ===")
			fmt.Printf("Всего задач: %d\n", stats["total"])
			fmt.Printf("Выполнено: %d\n", stats["done"])
			fmt.Printf("Осталось: %d\n", stats["pending"])
			quadrants := stats["quadrants"].(map[string]int)
			for q, count := range quadrants {
				fmt.Printf("  %s: %d\n", q, count)
			}
		case "7":
			if err := matrix.SaveToFile("eisenhower_data.json"); err != nil {
				fmt.Println("Ошибка сохранения:", err)
			} else {
				fmt.Println("Сохранено.")
			}
		case "8":
			if err := matrix.LoadFromFile("eisenhower_data.json"); err != nil {
				fmt.Println("Ошибка загрузки:", err)
			} else {
				fmt.Println("Загружено.")
			}
		default:
			fmt.Println("Неизвестная команда.")
		}
	}
}

File: Matrix.java
// Matrix.java
import java.io.*;
import java.nio.file.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

enum Quadrant {
    DO("Важно и срочно"),
    PLAN("Важно, но не срочно"),
    DELEGATE("Не важно, но срочно"),
    ELIMINATE("Не важно и не срочно");

    private final String display;
    Quadrant(String display) { this.display = display; }
    public String getDisplay() { return display; }
}

record Task(int id, String title, String description, boolean important, boolean urgent, boolean done, String createdAt) implements Serializable {
    public Quadrant quadrant() {
        if (important && urgent) return Quadrant.DO;
        if (important) return Quadrant.PLAN;
        if (urgent) return Quadrant.DELEGATE;
        return Quadrant.ELIMINATE;
    }
}

class EisenhowerMatrix implements Serializable {
    private static final long serialVersionUID = 1L;
    private List<Task> tasks = new ArrayList<>();
    private int nextId = 1;

    public Task addTask(String title, String desc, boolean important, boolean urgent) {
        Task task = new Task(nextId, title, desc, important, urgent, false,
                LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME));
        tasks.add(task);
        nextId++;
        return task;
    }

    public Optional<Task> findTask(int id) {
        return tasks.stream().filter(t -> t.id() == id).findFirst();
    }

    public boolean editTask(int id, String title, String desc, Boolean important, Boolean urgent) {
        Optional<Task> opt = findTask(id);
        if (opt.isEmpty()) return false;
        Task old = opt.get();
        tasks.remove(old);
        Task updated = new Task(
                old.id(),
                title != null ? title : old.title(),
                desc != null ? desc : old.description(),
                important != null ? important : old.important(),
                urgent != null ? urgent : old.urgent(),
                old.done(),
                old.createdAt()
        );
        tasks.add(updated);
        return true;
    }

    public boolean toggleDone(int id) {
        Optional<Task> opt = findTask(id);
        if (opt.isEmpty()) return false;
        Task old = opt.get();
        tasks.remove(old);
        Task updated = new Task(
                old.id(), old.title(), old.description(),
                old.important(), old.urgent(), !old.done(), old.createdAt()
        );
        tasks.add(updated);
        return true;
    }

    public boolean deleteTask(int id) {
        return tasks.removeIf(t -> t.id() == id);
    }

    public List<Task> filterByQuadrant(Quadrant q) {
        return tasks.stream().filter(t -> t.quadrant() == q).collect(Collectors.toList());
    }

    public Map<String, Object> getStats() {
        int total = tasks.size();
        int done = (int) tasks.stream().filter(Task::done).count();
        Map<Quadrant, Long> quadCounts = tasks.stream().collect(Collectors.groupingBy(Task::quadrant, Collectors.counting()));
        Map<String, Object> stats = new HashMap<>();
        stats.put("total", total);
        stats.put("done", done);
        stats.put("pending", total - done);
        stats.put("quadrants", quadCounts);
        return stats;
    }

    public void saveToFile(String filename) throws IOException {
        try (ObjectOutputStream oos = new ObjectOutputStream(Files.newOutputStream(Paths.get(filename)))) {
            oos.writeObject(this);
        }
    }

    public void loadFromFile(String filename) throws IOException, ClassNotFoundException {
        try (ObjectInputStream ois = new ObjectInputStream(Files.newInputStream(Paths.get(filename)))) {
            EisenhowerMatrix loaded = (EisenhowerMatrix) ois.readObject();
            this.tasks = loaded.tasks;
            this.nextId = loaded.nextId;
        }
    }

    public List<Task> getTasks() { return Collections.unmodifiableList(tasks); }
}

public class Matrix {
    private static final Scanner scanner = new Scanner(System.in);

    private static String readString(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private static boolean readBool(String prompt) {
        while (true) {
            String input = readString(prompt);
            if (input.isEmpty()) return false;
            if (input.equalsIgnoreCase("y")) return true;
            if (input.equalsIgnoreCase("n")) return false;
            System.out.println("Введите y или n.");
        }
    }

    private static int readInt(String prompt) {
        while (true) {
            try {
                System.out.print(prompt);
                return Integer.parseInt(scanner.nextLine().trim());
            } catch (NumberFormatException e) {
                System.out.println("Введите число.");
            }
        }
    }

    private static void printTask(Task task, boolean showQuadrant) {
        String doneMark = task.done() ? "✅" : "⬜";
        System.out.printf("%s #%d - %s%n", doneMark, task.id(), task.title());
        System.out.printf("   Описание: %s%n", task.description());
        if (showQuadrant) {
            System.out.printf("   Квадрант: %s (важно: %b, срочно: %b)%n",
                    task.quadrant().getDisplay(), task.important(), task.urgent());
        }
        System.out.printf("   Создана: %s%n", task.createdAt());
    }

    public static void main(String[] args) {
        EisenhowerMatrix matrix = new EisenhowerMatrix();
        try {
            matrix.loadFromFile("eisenhower_data.ser");
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Не удалось загрузить данные, начинаем с пустой доски.");
        }

        while (true) {
            System.out.println("\n===== МАТРИЦА ЭЙЗЕНХАУЭРА (Java) =====");
            System.out.println("1. Показать все задачи");
            System.out.println("2. Добавить задачу");
            System.out.println("3. Редактировать задачу");
            System.out.println("4. Отметить выполненной / невыполненной");
            System.out.println("5. Удалить задачу");
            System.out.println("6. Показать статистику");
            System.out.println("7. Сохранить в файл");
            System.out.println("8. Загрузить из файла");
            System.out.println("0. Выход");
            String choice = readString("Выберите действие: ");

            switch (choice) {
                case "0" -> { return; }
                case "1" -> {
                    if (matrix.getTasks().isEmpty()) {
                        System.out.println("Задач нет.");
                    } else {
                        for (Quadrant q : Quadrant.values()) {
                            var tasks = matrix.filterByQuadrant(q);
                            if (!tasks.isEmpty()) {
                                System.out.println("\n--- " + q.getDisplay() + " ---");
                                tasks.forEach(t -> printTask(t, false));
                            }
                        }
                    }
                }
                case "2" -> {
                    String title = readString("Введите название: ");
                    if (title.isBlank()) {
                        System.out.println("Название не может быть пустым.");
                        continue;
                    }
                    String desc = readString("Введите описание: ");
                    boolean important = readBool("Задача важна? (y/n): ");
                    boolean urgent = readBool("Задача срочна? (y/n): ");
                    Task task = matrix.addTask(title, desc, important, urgent);
                    System.out.printf("Задача добавлена с ID %d, квадрант: %s%n", task.id(), task.quadrant().getDisplay());
                }
                case "3" -> {
                    int id = readInt("Введите ID задачи: ");
                    var opt = matrix.findTask(id);
                    if (opt.isEmpty()) {
                        System.out.println("Задача не найдена.");
                        continue;
                    }
                    Task old = opt.get();
                    System.out.println("Оставьте поле пустым, чтобы не менять.");
                    String newTitle = readString("Новое название (" + old.title() + "): ");
                    String newDesc = readString("Новое описание (" + old.description() + "): ");
                    Boolean newImp = null;
                    String impInput = readString("Важна? (y/n) сейчас: " + (old.important() ? "y" : "n") + ": ");
                    if (!impInput.isEmpty()) newImp = impInput.equalsIgnoreCase("y");
                    Boolean newUrg = null;
                    String urgInput = readString("Срочна? (y/n) сейчас: " + (old.urgent() ? "y" : "n") + ": ");
                    if (!urgInput.isEmpty()) newUrg = urgInput.equalsIgnoreCase("y");
                    if (matrix.editTask(id,
                            newTitle.isEmpty() ? null : newTitle,
                            newDesc.isEmpty() ? null : newDesc,
                            newImp, newUrg)) {
                        System.out.println("Задача обновлена.");
                    } else {
                        System.out.println("Ошибка.");
                    }
                }
                case "4" -> {
                    int id = readInt("Введите ID задачи: ");
                    if (matrix.toggleDone(id)) {
                        System.out.println("Статус выполнения изменён.");
                    } else {
                        System.out.println("Задача не найдена.");
                    }
                }
                case "5" -> {
                    int id = readInt("Введите ID задачи для удаления: ");
                    if (matrix.deleteTask(id)) {
                        System.out.println("Задача удалена.");
                    } else {
                        System.out.println("Задача не найдена.");
                    }
                }
                case "6" -> {
                    var stats = matrix.getStats();
                    System.out.println("\n=== СТАТИСТИКА ===");
                    System.out.println("Всего задач: " + stats.get("total"));
                    System.out.println("Выполнено: " + stats.get("done"));
                    System.out.println("Осталось: " + stats.get("pending"));
                    @SuppressWarnings("unchecked")
                    Map<Quadrant, Long> quad = (Map<Quadrant, Long>) stats.get("quadrants");
                    for (var entry : quad.entrySet()) {
                        System.out.println("  " + entry.getKey().getDisplay() + ": " + entry.getValue());
                    }
                }
                case "7" -> {
                    try {
                        matrix.saveToFile("eisenhower_data.ser");
                        System.out.println("Сохранено.");
                    } catch (IOException e) {
                        System.out.println("Ошибка сохранения: " + e.getMessage());
                    }
                }
                case "8" -> {
                    try {
                        matrix.loadFromFile("eisenhower_data.ser");
                        System.out.println("Загружено.");
                    } catch (IOException | ClassNotFoundException e) {
                        System.out.println("Ошибка загрузки: " + e.getMessage());
                    }
                }
                default -> System.out.println("Неизвестная команда.");
            }
        }
    }
}
